mlir-sycl/lib/Dialect/SYCL/Transforms/ConstantPropagation.cpp

```diff
@@ -91,12 +112,630 @@ class ConstantArithArg : public ConstantExplicitArg {
+      .Case<ConstantSYCLGridArgs>(
+          [=](const auto *propagator) { return propagator->match(op); });
+}
+
+void ConstantImplicitArgBase::rewrite(Operation *op, OpBuilder &builder) const {
+  TypeSwitch<const ConstantImplicitArgBase *>(this).Case<ConstantSYCLGridArgs>(
+      [&](const auto *propagator) { return propagator->rewrite(op, builder); });
+}
+
+bool ConstantImplicitArgBase::isRecursive() const {
+  return TypeSwitch<const ConstantImplicitArgBase *, bool>(this)
+      .Case<ConstantSYCLGridArgs>(
+          [](const auto *ptr) { return ptr->isRecursive(); });
+}
+
+void ConstantImplicitArgBase::printDebugMessage() const {
+  TypeSwitch<const ConstantImplicitArgBase *>(this).Case<ConstantSYCLGridArgs>(
+      [](const auto *ptr) { return ptr->printDebugMessage(); });
+}
+
+void ConstantImplicitArgBase::propagate(FunctionOpInterface func) {
+  LLVM_DEBUG(printDebugMessage());
+  SymbolTableCollection symbolTable;
+  propagate(func, symbolTable, isRecursive());
+}
+
+bool ConstantImplicitArgBase::needsChanges(FunctionOpInterface func,
+                                           SymbolTableCollection &symbolTable,
+                                           bool recursive) const {
+  // Declaration
+  if (func.isExternal())
+    return false;
+  return func
+      .walk([&](Operation *op) {
+        if (succeeded(match(op)))
+          return WalkResult::interrupt();
+        if (!recursive)
+          return WalkResult::advance();
+        auto call = dyn_cast<CallOpInterface>(op);
+        if (!call)
+          return WalkResult::advance();
+        SymbolRefAttr callable =
+            call.getCallableForCallee().dyn_cast<SymbolRefAttr>();
+        if (!callable)
+          return WalkResult::advance();
+        auto called = symbolTable.lookupNearestSymbolFrom<FunctionOpInterface>(
+            call, callable);
+        assert(called && "Function not found");
+        return needsChanges(called, symbolTable, recursive)
+                   ? WalkResult::interrupt()
+                   : WalkResult::advance();
+      })
+      .wasInterrupted();
+}
+
+void ConstantImplicitArgBase::propagate(FunctionOpInterface func,
+                                        SymbolTableCollection &symbolTable,
+                                        bool recursive, Twine callPath,
+                                        CallOpInterface call) {
+  LLVM_DEBUG(llvm::dbgs().indent(6)
+             << "Entering function '"
+             << static_cast<SymbolOpInterface>(func).getName() << "'\n");
+  // Declaration
+  if (func.isExternal())
+    return;
+  StringRef originalName = static_cast<SymbolOpInterface>(func).getName();
+  // Need to clone called functions to specialize
+  if (call) {
+    // Avoid cloning if no change will be performed.
+    // Only check whether func needs changes here to avoid iterating twice.
+    if (!needsChanges(func, symbolTable, recursive))
+      return;
+    func = cloneFunction(func, call, callPath);
```

Comment:
is there situations that we can avoid cloning? like the function that need changes is private or has linkonce_odr linkage?

Reply:
Yes, there might be some situations. I'm thinking on implementing those optimizations in the future, but, due to time constraints, I'm taking the shortcut here of safely assuming we have to clone. If the original one is not needed and can be removed, it will be removed in further stages anyway.